File: packages/julia/src/Seguranca/Ataques/AtaqueSAC.jl
"""
SAC (Strict Avalanche Criterion): para CADA bit de entrada (IV), ao virar esse
bit, CADA bit de saída deve mudar com probabilidade ~0.5. Mede o pior bit de
saída; se algum fica longe de 50%, a difusão tem ponto cego.
"""
struct AtaqueSAC <: Ataque
    tamanho_bloco::Int
    amostras::Int
    tolerancia::Float64
end

AtaqueSAC(;
    tamanho_bloco::Integer = 32,
    amostras::Integer = 40,
    tolerancia::Real = 0.05,
) = AtaqueSAC(Int(tamanho_bloco), Int(amostras), Float64(tolerancia))

nome(::AtaqueSAC) = "SAC (Strict Avalanche Criterion)"

function executar(a::AtaqueSAC, alvo::AlvoCriptografico)
    chave = to_buffer(chave_de_teste(alvo))
    total_bits = a.tamanho_bloco * 8
    flips = zeros(Int, total_bits)
    tam_iv = tamanho_iv(alvo)
    total = 0

    for pos in 0:(tam_iv-1)
        for bit in 0:7
            for _ in 1:(a.amostras)
                iv1 = random_bytes(tam_iv)
                iv2 = copy(iv1)
                iv2[pos+1] = iv2[pos+1] ⊻ UInt8(1 << bit)

                ks1 = gerar_keystream_bruto(alvo, chave, iv1, "enc", a.tamanho_bloco)
                ks2 = gerar_keystream_bruto(alvo, chave, iv2, "enc", a.tamanho_bloco)

                for j in 0:(total_bits-1)
                    b1 = (ks1[div(j, 8)+1] >> (7 - (j & 7))) & 0x01
                    b2 = (ks2[div(j, 8)+1] >> (7 - (j & 7))) & 0x01
                    if b1 != b2
                        flips[j+1] += 1
                    end
                end
                total += 1
            end
        end
    end

    pior = -1
    pior_desvio = 0.0
    pior_p = 0.5
    for j in 0:(total_bits-1)
        p = flips[j+1] / total
        desvio = abs(p - 0.5)
        if desvio > pior_desvio
            pior_desvio = desvio
            pior = j
            pior_p = p
        end
    end

    vulneravel = pior_desvio > a.tolerancia

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "alta" : "info",
        "pior bit de saída=$pior: p=$(fmt(pior_p * 100, 2))% (esperado 50%, tolerância ±$(fmt(a.tolerancia * 100, 1))%); $total amostras por bit de entrada",
        Dict{String,Any}("pior" => pior, "pior_p" => pior_p, "desvio" => pior_desvio),
    )
end

File: packages/php/src/Seguranca/Ataques/AtaqueSAC.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * SAC (Strict Avalanche Criterion): para CADA bit de entrada (IV), ao virar
 * esse bit, CADA bit de saída deve mudar com probabilidade ~0.5. Mede o pior
 * bit de saída; se algum fica longe de 50%, a difusão tem ponto cego.
 */
class AtaqueSAC implements AtaqueInterface
{
    public function __construct(
        private int $tamanhoBloco = 32,
        private int $amostras = 40,
        private float $tolerancia = 0.05,
    ) {
    }

    public function nome(): string
    {
        return 'SAC (Strict Avalanche Criterion)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $chave = $alvo->chaveDeTeste();
        $totalBits = $this->tamanhoBloco * 8;
        $flips = array_fill(0, $totalBits, 0);
        $tamanhoIv = $alvo->tamanhoIv();
        $total = 0;

        $primeiro = $alvo->gerarKeystreamBruto($chave, random_bytes($tamanhoIv), 'enc', $this->tamanhoBloco);
        if ($primeiro === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        for ($pos = 0; $pos < $tamanhoIv; $pos++) {
            for ($bit = 0; $bit < 8; $bit++) {
                for ($s = 0; $s < $this->amostras; $s++) {
                    $iv1 = random_bytes($tamanhoIv);
                    $iv2 = $iv1;
                    $iv2[$pos] = chr(ord($iv2[$pos]) ^ (1 << $bit));

                    $ks1 = $alvo->gerarKeystreamBruto($chave, $iv1, 'enc', $this->tamanhoBloco);
                    $ks2 = $alvo->gerarKeystreamBruto($chave, $iv2, 'enc', $this->tamanhoBloco);

                    for ($j = 0; $j < $totalBits; $j++) {
                        $b1 = (ord($ks1[$j >> 3]) >> (7 - ($j & 7))) & 1;
                        $b2 = (ord($ks2[$j >> 3]) >> (7 - ($j & 7))) & 1;
                        if ($b1 !== $b2) {
                            $flips[$j]++;
                        }
                    }
                    $total++;
                }
            }
        }

        $pior = -1;
        $piorDesvio = 0.0;
        $piorP = 0.5;
        for ($j = 0; $j < $totalBits; $j++) {
            $p = $flips[$j] / $total;
            $desvio = abs($p - 0.5);
            if ($desvio > $piorDesvio) {
                $piorDesvio = $desvio;
                $pior = $j;
                $piorP = $p;
            }
        }

        $vulneravel = $piorDesvio > $this->tolerancia;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'alta' : 'info',
            detalhes: sprintf(
                'pior bit de saída=%d: p=%.2f%% (esperado 50%%, tolerância ±%.1f%%); %d amostras por bit de entrada',
                $pior,
                $piorP * 100,
                $this->tolerancia * 100,
                $total,
            ),
            dados: ['pior' => $pior, 'pior_p' => $piorP, 'desvio' => $piorDesvio],
        );
    }
}
